pkg/solana/logpoller/parser.go

```diff
@@ -1,5 +1,413 @@
+	txHashFieldName    = "tx_hash"
+	addressFieldName   = "address"
+	eventSigFieldName  = "event_sig"
+	defaultSort        = "block_number ASC, log_index ASC"
+)
+
 var (
-	logsFields = [...]string{"id", "filter_id", "chain_id", "log_index", "block_hash", "block_number", "block_timestamp", "address", "event_sig", "subkey_values", "tx_hash", "data", "created_at", "expires_at", "sequence_num"}
+	ErrInvalidComparator   = errors.New("invalid comparison operator")
+	ErrInvalidConfidence   = errors.New("invalid confidence level; solana only supports finalized")
+	ErrInvalidCursorDir    = errors.New("invalid cursor direction")
+	ErrInvalidCursorFormat = errors.New("invalid cursor format")
+	ErrInvalidSortDir      = errors.New("invalid sort direction")
+	ErrInvalidSortType     = errors.New("invalid sort by type")
+
+	logsFields = [...]string{"id", "filter_id", "chain_id", "log_index", "block_hash", "block_number",
+		"block_timestamp", "address", "event_sig", "subkey_values", "tx_hash", "data", "created_at",
+		"expires_at", "sequence_num"}
 )
+
+// The parser builds SQL expressions piece by piece for each Accept function call and resets the error and expression
+// values after every call.
+type pgDSLParser struct {
+	args *queryArgs
+
+	// transient properties expected to be set and reset with every expression
+	expression string
+	err        error
+}
+
+var _ primitives.Visitor = (*pgDSLParser)(nil)
+
+func (v *pgDSLParser) Comparator(_ primitives.Comparator) {}
+
+func (v *pgDSLParser) Block(prim primitives.Block) {
+	cmp, err := cmpOpToString(prim.Operator)
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	v.expression = fmt.Sprintf(
+		"%s %s :%s",
+		blockFieldName,
+		cmp,
+		v.args.withIndexedField(blockFieldName, prim.Block),
+	)
+}
+
+func (v *pgDSLParser) Confidence(prim primitives.Confidence) {
+	switch prim.ConfidenceLevel {
+	case primitives.Finalized, primitives.Unconfirmed:
+		// solana LogPoller will only use and store finalized logs
+		// to ensure x-chain compatibility, do nothing and return no error
+		// confidence in solana is effectively a noop
+		return
+	default:
+		// still return an error for invalid confidence levels
+		v.err = fmt.Errorf("%w: %s", ErrInvalidConfidence, prim.ConfidenceLevel)
+
+		return
+	}
+}
+
+func (v *pgDSLParser) Timestamp(prim primitives.Timestamp) {
+	cmp, err := cmpOpToString(prim.Operator)
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	var tm int64
+	if prim.Timestamp > math.MaxInt64 {
+		tm = 0
+	} else {
+		tm = int64(prim.Timestamp)
+	}
+
+	v.expression = fmt.Sprintf(
+		"%s %s :%s",
+		timestampFieldName,
+		cmp,
+		v.args.withIndexedField(timestampFieldName, time.Unix(tm, 0)),
+	)
+}
+
+func (v *pgDSLParser) TxHash(prim primitives.TxHash) {
+	txHash, err := solana.PublicKeyFromBase58(prim.TxHash)
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	v.expression = fmt.Sprintf(
+		"%s = :%s",
+		txHashFieldName,
+		v.args.withIndexedField(txHashFieldName, txHash),
```

Comment:
solana.PublicKey does not implement Value method, so not sure if the query won't error. Might need to wrap it into type defined in `types` file.